File: WebAPI/Services/ProductsService.cs
using Models;
using Services.DTO;
using Services.Interfaces;
using System.Linq;

namespace Services
{
    public class ProductsService : IProductsService
    {
        private readonly Database database;

        public ProductsService(Database database)
        {
            this.database = database;
        }
        public bool Delete(int productId)
        {
            Product product = this.database.Products
                .Where(p => p.Id == productId)
                .FirstOrDefault();

            if (product == null)
                return false;

            var basketItems = this.database.BasketItems
                .Where(bi => bi.ProductId == productId)
                .AsEnumerable();

            this.database.BasketItems.RemoveRange(basketItems);
            this.database.Products.Remove(product);
            this.database.SaveChanges();
            return true;
        }

        public PaginatedData<ProductDto> Get(PaginationDto dto)
        {
            PaginatedData<ProductDto> paginatedData = new PaginatedData<ProductDto>();
            paginatedData.Data = this.database.Products
                .Select(p => new ProductDto { 
                    Id = p.Id,
                    Name = p.Name,
                    Description = p.Description,
                    Price = p.Price
                });
            paginatedData.Count = paginatedData.Data.Count();

            switch(dto.SortColumn.ToLower())
            {
                case "name":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(p => p.Name);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(p => p.Name);
                        break;
                    }
                case "description":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(p => p.Description);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(p => p.Description);
                        break;
                    }
                case "price":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(p => p.Price);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(p => p.Price);
                        break;
                    }
            }

            paginatedData.Data = paginatedData.Data.Skip((dto.Page - 1) * dto.RowsPerPage).Take(dto.RowsPerPage);
            return paginatedData;
        }

        public ProductDto GetById(int productId)
        {
            Product product = this.database.Products
                .Where(p => p.Id == productId)
                .FirstOrDefault();

            if (product == null)
                return null;

            return new ProductDto
            {
                Id = product.Id,
                Name = product.Name,
                Description = product.Description,
                Price = product.Price
            };
        }

        public ProductDto Post(PostProductDto dto)
        {
            Product product = new Product()
            {
                Name = dto.Name,
                Description = dto.Description,
                Price = dto.Price
            };

            this.database.Products.Add(product);
            this.database.SaveChanges();

            return new ProductDto
            {
                Id = product.Id,
                Name = product.Name,
                Description = product.Description,
                Price = product.Price
            };
        }

        public ProductDto Put(int productId, PutProductDto dto)
        {
            Product product = this.database.Products
                .Where(p => p.Id == productId)
                .FirstOrDefault();

            if (product == null)
                return null;

            product.Name = dto.Name;
            product.Description = dto.Description;
            product.Price = dto.Price;
            this.database.SaveChanges();

            return new ProductDto
            {
                Id = product.Id,
                Name = product.Name,
                Description = product.Description,
                Price = product.Price
            };
        }
    }
}

File: website/src/app/services/products.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { PaginatedData } from '../models/paginated-data';
import { Pagination } from '../models/pagination';
import { PostProduct } from '../models/post-product';
import { Product } from '../models/product';
import { PutProduct } from '../models/put-product';

@Injectable({
  providedIn: 'root'
})
export class ProductsService {
  constructor(private httpClient: HttpClient) { }

  get(pagination?: Pagination): Observable<PaginatedData<Product>> {
    return this.httpClient
          .get<PaginatedData<Product>>('http://localhost:1234/Products?'
          + 'SortColumn=' + (pagination?.sortColumn ?? 'name')
          + '&Page=' + (pagination?.page ?? 1)
          + '&RowsPerPage=' + (pagination?.rowsPerPage ?? 10)
          + '&SortAscending=' + (pagination?.sortAscending ?? true));
  }

  getById(productId: number): Observable<Product> {
    return this.httpClient.get<Product>('http://localhost:1234/Products/' + productId);
  }

  put(productId: number, dto: PutProduct): Observable<Product> {
    return this.httpClient.put<Product>('http://localhost:1234/Products/' + productId, dto)
  }

  post(dto: PostProduct): Observable<Product> {
    return this.httpClient.post<Product>('http://localhost:1234/Products', dto)
  }

  delete(productId: number): Observable<boolean> {
    return this.httpClient.delete<boolean>('http://localhost:1234/Products/' + productId);
  }
}
